pkg/runtime/wfengine/wfengine.go

```diff
@@ -145,41 +148,40 @@ func (wfe *WorkflowEngine) SetActorReminderInterval(interval time.Duration) {
 }
 
 func (wfe *WorkflowEngine) Start(ctx context.Context) error {
-	var err error
-	wfe.startedOnce.Do(func() {
-		if wfe.actorRuntime == nil {
-			err = errors.New("actor runtime is not configured")
-			return
-		} else if wfe.executor == nil {
-			err = errors.New("grpc executor is not yet configured")
-			return
-		}
+	// Start could theoretically get called by multiple goroutines concurrently
+	wfe.startMutex.Lock()
+	defer wfe.startMutex.Unlock()
 
-		for actorType, actor := range wfe.InternalActors() {
-			err = wfe.actorRuntime.RegisterInternalActor(ctx, actorType, actor)
-			if err != nil {
-				err = fmt.Errorf("failed to register workflow actor %s: %w", actorType, err)
-				return
-			}
-		}
+	if wfe.IsRunning {
+		return nil
+	}
 
-		// TODO: Determine whether a more dynamic parallelism configuration is necessary.
-		parallelismOpts := backend.WithMaxParallelism(100)
+	if wfe.actorRuntime == nil {
+		return errors.New("actor runtime is not configured")
+	} else if wfe.executor == nil {
+		return errors.New("grpc executor is not yet configured")
+	}
 
-		orchestrationWorker := backend.NewOrchestrationWorker(wfe.backend, wfe.executor, wfLogger, parallelismOpts)
-		activityWorker := backend.NewActivityTaskWorker(wfe.backend, wfe.executor, wfLogger, parallelismOpts)
-		wfe.worker = backend.NewTaskHubWorker(wfe.backend, orchestrationWorker, activityWorker, wfLogger)
-		err = wfe.worker.Start(ctx)
-		if err != nil {
-			err = fmt.Errorf("failed to start workflow engine: %w", err)
-			return
+	for actorType, actor := range wfe.InternalActors() {
+		if err := wfe.actorRuntime.RegisterInternalActor(ctx, actorType, actor); err != nil {
+			return fmt.Errorf("failed to register workflow actor %s: %w", actorType, err)
 		}
+	}
 
-		wfe.IsRunning = true
-		wfLogger.Info("workflow engine started")
-	})
+	// TODO: Determine whether a more dynamic parallelism configuration is necessary.
+	parallelismOpts := backend.WithMaxParallelism(100)
+
+	orchestrationWorker := backend.NewOrchestrationWorker(wfe.backend, wfe.executor, wfLogger, parallelismOpts)
+	activityWorker := backend.NewActivityTaskWorker(wfe.backend, wfe.executor, wfLogger, parallelismOpts)
+	wfe.worker = backend.NewTaskHubWorker(wfe.backend, orchestrationWorker, activityWorker, wfLogger)
+	if err := wfe.worker.Start(ctx); err != nil {
+		return fmt.Errorf("failed to start workflow engine: %w", err)
+	}
 
-	return err
+	wfe.IsRunning = true
+	wfLogger.Info("workflow engine started")
+
+	return nil
```

Comment:
(comment about L197, I can't comment there)

The method `Stop()` doesn't have a lock so there could be race conditions in modifying `IsRunning`